### ISSUE TITLE
"Server is Starting" when using Phusion Passenger

### ISSUE THREAD

**Issue by frnco** [krakenjs/kraken-js#326]
This piece of code breaks flows with Phusion Passenger, since Passenger starts processes when requests are made:

        parent.use(function startup(req, res, next) {
            if (deferred.promise.isFulfilled()) {
                next();
                return;
            }z
            res.send(503, 'Server is starting.');
        });

Wouldn't it be better to check on promises and wait until they complete if needed?

Not the easiest way out, I'm pretty sure, but I think it's better for KrakenJS if it can run amazingly with lots of tools, and Passenger is a pretty good tool for deloyment.

Also, not directly related and I know off-topic is kinda bad, but I couldn't find much about deploying. I'll possibly create another issue or do a Pull Request on the docs about that (So please let's not fill this thread with deploying stuff), but it won't hurt to ask opinions on the matter, maybe I just need some pointing in the right direction.

Thanks in advance. :D

**Comment by totherik** [comment]
I'll let @jasisk decide the future direction, but I can say that blocking/holding onto requests was taken into account when designing this. I decided that it was inappropriate to queue an indeterminate number of requests and instead allow infrastructure designed to understand application readiness do their job. 

I also don't want to rabbit hole on deployment concerns, but a running process or bound port do not an application make. The application should dictate when it's able to handle requests and behave accordingly. 

Interested to hear @jasisk thoughts.

**Comment by frnco** [comment]
@totherik Yes, I found some older discussion on that matter and although it was somewhat out of my focus at the time (Specifically looking for why I was receiving a message) I got the idea that KrakenJS would "fail instataneously" instead of waiting on listens.

Obviously this has some advantages. I'm just curious about how we should deal when we need many instances (Or generate new instances on the fly), since we stop being able to use tools like passenger, and about what would be the "recommended way" to deploy an application, since there may be other perks I'm unaware of.


And by the way, you guys have done/are doing an amazing job with KrakenJS. It's amazing we already have something this polished on NodeJS.

**Comment by jasisk** [comment]
Hey @frnco. Thanks for the kind words. We really appreciate it.

If you look just above the block of code you mention, you'll find that when the promise [is resolved](https://github.com/krakenjs/kraken-js/blob/eb84151171fea26f3a8a0519724f98d7b83d2307/index.js#L68), we [emit a `start` event](https://github.com/krakenjs/kraken-js/blob/eb84151171fea26f3a8a0519724f98d7b83d2307/index.js#L61). You can defer binding until that point such that there will never be a case when the promise in unfulfilled and bound.

Since Passenger hooks that first call to listen, I believe that will take care of the issue. I'm unable to run Passenger at the moment, but give something like this a shot:

``` js
'use strict';
var kraken = require('kraken-js');
var express = require('express');
var app = express();

app.use(kraken());

app.on('start', function () {
  app.listen(3000, function () {
    console.log('listening ...');
  });
});
```

**Comment by frnco** [comment]
@jasisk That woked like a charm. Now the only problem I have with Passenger is the CSRF, but for this project I think I can safely disable it, and this is something I should check about on lusca anyways.

Nevertheless, I get the idea that the intended flow for KrakenJS is different from the one I'm using. Not that I'm surprised, I'm mostly just curious about how you guys are deploying, and why documentation on deployment is so scarce.

**Comment by jasisk** [comment]
@frnco sorry for the delay—I must have missed your last comment.

Our (PayPal's) deployment process is pretty unique to our infrastructure which is why we don't talk about it too much. That said, we certainly have opinions on deployment outside of our Big Machine™ and can distill some aspects of it into reusable pieces of knowledge. Something we should keep in mind for the near term but wouldn't be for a little while.

Curious what solutions you end up with. Regardless, best of luck!

**Comment by frnco** [comment]
@jasisk Init scripts seem to be the way for simple stuff (without Upstart since it's pretty much dead), but for more complex systems I'm using Redis for sessions. That makes it work perfectly well, plus it helps a lot if the software needs to scale, and ultimately seems to be the better solution.

**Comment by jasisk** [comment]
Yup. Agree completely. An out of band process manager is a must. Same with an out of band session store (not bound to the process!).
